pyuvdata/uvcal/ms_cal.py

```diff
@@ -0,0 +1,582 @@
+        ant_dict = {ant: idx for idx, ant in enumerate(self.antenna_numbers)}
+        cal_arr_shape = (self.Nants_data, nchan, self.Ntimes, self.Njones)
+
+        ms_cal_soln = np.zeros(
+            cal_arr_shape, dtype=complex if (self.cal_type == "gain") else float
+        )
+        self.quality_array = np.zeros(cal_arr_shape, dtype=float)
+        self.flag_array = np.ones(cal_arr_shape, dtype=bool)
+        self.total_quality_array = None  # Always None for now, no similar array in MS
+        self.scan_number_array = np.zeros_like(self.time_array, dtype=int)
+        self.phase_center_id_array = np.zeros_like(self.time_array, dtype=int)
+
+        for row_idx, time_idx in enumerate(row_timeidx_map):
+            try:
+                ant_idx = ant_dict[tb_main.getcell("ANTENNA1", row_idx)]
+                time_val = tb_main.getcell("TIME", row_idx)
+                cal_soln = tb_main.getcell(cal_column, row_idx)
+                cal_qual = tb_main.getcell("PARAMERR", row_idx)
+                cal_flag = tb_main.getcell("FLAG", row_idx)
+                field_id = tb_main.getcell("FIELD_ID", row_idx)
+                scan_num = tb_main.getcell("SCAN_NUMBER", row_idx)
+                int_time = tb_main.getcell("INTERVAL", row_idx)
+                spw_id = tb_main.getcell("SPECTRAL_WINDOW_ID", row_idx)
+
+                # Figure out which spectral slice this corresponds to
+                spw_slice = spw_slice_dict[spw_id]
+
+                # Finally, start plugging in solns to various parameters.
+                ms_cal_soln[ant_idx, spw_slice, time_idx, :] = cal_soln
+                self.time_array[time_idx] = time_val
+                self.integration_time[time_idx] = int_time
+                self.quality_array[ant_idx, spw_slice, time_idx, :] = cal_qual
+                self.flag_array[ant_idx, spw_slice, time_idx, :] = cal_flag
+                self.phase_center_id_array[time_idx] = field_id
+                self.scan_number_array[time_idx] = scan_num
+            except KeyError:
+                # If there's no entry that matches, it's because we've effectively
+                # flagged some index value such that it has no entries in the table.
+                # skip recording this row.
+                continue
+
+        # Convert the time from MJD secs (CASA standard) to JD date (pyuvdata std)
+        self.time_array = Time(
+            self.time_array / 86400.0,
+            format="mjd",
+            scale=ms_utils._get_time_scale(tb_main),
+        ).utc.jd
+
+        # There's a little bit of cleanup to do w/ field_id, since the values here
+        # correspond to the row in FIELD rather than the source ID. Map that now.
+        field_id_map = dict(enumerate(self.phase_center_catalog))
+        self.phase_center_id_array = np.array(
+            [field_id_map[idx] for idx in self.phase_center_id_array]
+        )
+
+        # I think this is always east.
+        self.x_orientation = "east"
```

Comment:
Not that I've seen thusfar -- gains solutions appear to be a bit weird in that they don't have the full set of tables that say the visibility containers do (this is also a problem for tracking the Jones/polarization axes), I think it part because the gains tables are seen as something of an extension of a full MeasurementSet (even though they're not strictly tied to it). Obviously the above isn't appropriate for every telescope -- I just set it to allow me to march forward. Maybe we should be gathering some defaults and returning that as part of `UVTelescope.get_telescope`?